# Primal, Dual and Complementary Slackness

In [13]:
import numpy as np
import sympy
from sympy import *
from scipy.optimize import linprog

import warnings
warnings.filterwarnings('ignore')

# Optimization Problem

*PAR* is a small manufacturer of golf equipment and accessories whose distributor convinced them that there is a market for both the mid-priced golf bag, known as the standard model, and for a high-priced golf bag, known as the deluxe model. The distributor is so confident in the market that if *PAR* can manufacture the bags competitively, the distributor agrees to purchase all the bags *PAR* can produce over the next three months.

A careful analysis of the manufacturing requirements resulted in the following table, which shows the time and production needs for the four required manufacturing operations and the accounting department's estimate of the contribution to profit per bag:

| **Product** 	| **Cutting and Dyeing (h)** 	| **Sewing (h)** 	| **Finishing (h)** 	| **Inspection and Packaging (h)** 	| **Profit per bag ($)** 	|
|--------------	|-----------------------------	|-----------------	|-------------------	|----------------------------------	|-------------------------	|
| Standard     	| 7/10                        	| 1/2             	| 1                 	| 1/10                             	| 7                       	|
| Deluxe       	| 1                           	| 5/6             	| 2/3               	| 1/4                              	| 9                       	|

The manufacturing director estimates that during the next three months, there will be 630 hours of cutting and dyeing time, 600 hours of sewing time, 708 hours of finishing time, and 135 hours of inspection and packaging time available for the production of golf bags.

# Problem 1

#### If *PAR* wishes to maximize its profit, what amount of standard and deluxe bags should it manufacture?

Solve the problem using the primal and dual form of the optimization problem.

#### Original Optimization Problem Statement

Decision Variables:

$x_1 \longrightarrow $ Quantity of standard bags

$x_2 \longrightarrow $ Quantity of deluxe bags

Objective Function: Profit

$f(x) = 7x_1 + 9x_2$

Constraints: Time in hours

$\frac{7}{10}x_1 + x_2 \leq 630 $ 

$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ 

$x_1 + \frac{2}{3}x_2 \leq 708$

$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Original Optimization Problem: 

max $f(x) = 7x_1 + 9x_2$ s.t.

$\frac{7}{10}x_1 + x_2 \leq 630 $ 

$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ 

$x_1 + \frac{2}{3}x_2 \leq 708$ 

$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Primal Problem: 

min $-f(x) = -7x_1 - 9x_2$ s.t.

$-\frac{7}{10}x_1 - x_2 \geq -630 $ 

$-\frac{1}{2}x_1 - \frac{5}{6}x_2 \geq -600$ 

$-x_1 - \frac{2}{3}x_2 \geq -708$ 

$-\frac{1}{10}x_1 - \frac{1}{4}x_2 \geq -135$

Dual Problem:

max $q(\lambda) = -630\lambda_1 -600\lambda_2 - 708\lambda_3 - 135\lambda_4$ s.t. 

$-\frac{7}{10}\lambda_1 - \frac{1}{2}\lambda_2 - \lambda_3 - \frac{1}{10}\lambda_4 \leq -7$ 

$-\lambda_1 - \frac{5}{6}\lambda_2 - \frac{2}{3}\lambda_3 - \frac{1}{4}\lambda_4 \leq -9 $


In [14]:
c = np.array([[-7, -9]])
A = np.array([[7/10, 1],[1/2, 5/6],[1, 2/3],[1/10, 1/4]])
b = np.array([[630],[600],[708],[135]])

x1_bounds = (0, None)
x2_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds]), method='simplex', options={"disp": False})
print(f'The solution to the original optimization problem is x = {res.x}')
print(f'The value of the objective function evaluated at the optimal value is {-1*res.fun}')


The solution to the original optimization problem is x = [540. 252.]
The value of the objective function evaluated at the optimal value is 6048.0


According to this result, it is obtained that **540 standard bags and 252 deluxe bags** should be manufactured to maximize the profit to a value of **$6,048**.

The profit obtained by manufacturing **only standard bags** corresponds to **$4,956**.

The profit obtained by manufacturing **only deluxe bags** corresponds to **$4,860**.

# Problem 2

### a)
#### Planteamiento de la primera estrategia

Variables de decisión:

$x_1 \longrightarrow $ Cantidad de bolsas estándar \
$x_2 \longrightarrow $ Cantidad de bolsas deluxe

Función objetivo: Utilidad

$f(x) = 7x_1 + 18x_2$

Restricciones: Tiempo en horas

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Problema de optimización: \
max $f(x) = 7x_1 + 18x_2$ s.a.

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Problema primal: \
min $-f(x) = -7x_1 - 18x_2$ s.a.

$-\frac{7}{10}x_1 - x_2 \geq -630 $ \
$-\frac{1}{2}x_1 - \frac{5}{6}x_2 \geq -600$ \
$-x_1 - \frac{2}{3}x_2 \geq -708$ \
$-\frac{1}{10}x_1 - \frac{1}{4}x_2 \geq -135$

#### Planteamiento del problema dual de la primera estrategia

max $q(\lambda) = 630\lambda_1 + 600\lambda_2 + 708\lambda_3 + 135\lambda_4$ s.a. \
$\frac{7}{10}\lambda_1 + \frac{1}{2}\lambda_2 + \lambda_3 + \frac{1}{10}\lambda_4 \geq -7$ \
$\lambda_1 + \frac{5}{6}\lambda_2 + \frac{2}{3}\lambda_3 + \frac{1}{4}\lambda_4 \geq -18 $

#### Solución del problema dual de la primera estrategia

In [15]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[-7/10, -1/2, -1, -1/10],[-1, -5/6, -2/3, -1/4],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7],[-18],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex', options={"disp": True})
print(f'La solución a este problema de optimización es λ = {res.x}')
print(f'El valor de la función objetivo del dual evaluada en el valor óptimo es {res.fun}')

Optimization terminated successfully.
         Current function value: 9720.000000 
         Iterations: 3
La solución a este problema de optimización es λ = [ 0.  0.  0. 72.]
El valor de la función objetivo del dual evaluada en el valor óptimo es 9720.0


#### Solución del primal a través de holgura complementaria

A partir de la solución anterior, se evaluán las condiciones de holgura complementaria que permiten encontrar la solución en el problema primal, es decir, las cantidades de bolsas a fabricar de modelo estándar y modelo deluxe. Se obtienen las siguientes restricciones: 

$-\frac{7}{10}(0) - \frac{1}{2}(0) - 1(0) -\frac{1}{10}(72) < -7$\
$-7.2 < -7 \longrightarrow$ Restricción inactiva 

$-1(0) - \frac{5}{6}(0) - \frac{2}{3}(0) -\frac{1}{4}(72) < -18$
$-18 = -18 \longrightarrow$ Restricción activa 

$\lambda_1^*=0\longrightarrow$ Restricción activa 

$\lambda_2^*=0\longrightarrow$ Restricción activa 

$\lambda_3^*=0\longrightarrow$ Restricción activa 

$\lambda_4^*>0\longrightarrow$ Restricción inactiva 

De la primera y sexta restricción, se obtiene respectivamente que:

$x_1^*=0$ \
$-\frac{1}{10}x_1^*-\frac{1}{4}x_2^*=-135$

Finalmente, se resuelve el sistema y se obtiene que se deben fabricar 0 bolsas modelo estándar y 540 bolsas modelo deluxe.


### b) 
#### Planteamiento de la segunda estrategia

Variables de decisión:

$x_1 \longrightarrow $ Cantidad de bolsas estándar \
$x_2 \longrightarrow $ Cantidad de bolsas deluxe

Función objetivo: Utilidad

$f(x) = 14x_1 + 9x_2$

Restricciones: Tiempo en horas

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Problema de optimización: \
max $f(x) = 14x_1 + 9x_2$ s.a.

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

#### Planteamiento del problema dual de la segunda estrategia

max $g(\lambda) = 630\lambda_1 + 600\lambda_2 + 708\lambda_3 + 135\lambda_4$ s.a. \
$\frac{7}{10}\lambda_1 + \frac{1}{2}\lambda_2 + \lambda_3 + \frac{1}{10}\lambda_4 \geq -14$ \
$\lambda_1 + \frac{5}{6}\lambda_2 + \frac{2}{3}\lambda_3 + \frac{1}{4}\lambda_4 \geq -9 $

#### Solución del problema dual de la segunda estrategia

In [16]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[-7/10, -1/2, -1, -1/10],[-1, -5/6, -2/3, -1/4],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-14],[-9],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex')
print(f'La solución a este problema de optimización es λ = {res.x}')
print(f'El valor de la función objetivo del dual evaluada en el valor óptimo es {round(res.fun,3)}')

La solución a este problema de optimización es λ = [ 0.  0. 14.  0.]
El valor de la función objetivo del dual evaluada en el valor óptimo es 9912.0


#### Solución del primal a través de holgura complementaria

A partir de la solución anterior, se evaluán las condiciones de holgura complementaria que permiten encontrar la solución en el problema primal, es decir, las cantidades de bolsas a fabricar de modelo estándar y modelo deluxe. Se obtienen las siguientes restricciones: 

$-\frac{7}{10}(0) - \frac{1}{2}(0) - 1(14) -\frac{1}{10}(0) < -14$\
$-14 < -14 \longrightarrow$ Restricción activa 

$-1(0) - \frac{5}{6}(0) - \frac{2}{3}(14) -\frac{1}{4}(0) < -9$
$-9.33 = -9 \longrightarrow$ Restricción inactiva 

$\lambda_1^*=0\longrightarrow$ Restricción activa 

$\lambda_2^*=0\longrightarrow$ Restricción activa 

$\lambda_3^*>0\longrightarrow$ Restricción inactiva 

$\lambda_4^*=0\longrightarrow$ Restricción activa 

De la segunda y quinta restricción, se obtiene respectivamente que:

$x_2^*=0$ \
$-x_1^*-\frac{2}{3}x_2^*=-708$

Finalmente, se resuelve el sistema y se obtiene que se deben fabricar 708 bolsas modelo estándar y 0 bolsas modelo deluxe.

### c) 
#### Planteamiento de la tercera estrategia

Variables de decisión:

$x_1 \longrightarrow $ Cantidad de bolsas estándar \
$x_2 \longrightarrow $ Cantidad de bolsas deluxe

Función objetivo: Utilidad

$f(x) = 7x_1 + 9x_2$

Restricciones: Tiempo en horas

$\frac{7}{10}x_1 + x_2 \leq 1071 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 1020$ \
$x_1 + \frac{2}{3}x_2 \leq 1203.6$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 229.5$

Problema de optimización: \
max $f(x) = 7x_1 + 9x_2$ s.a.

$\frac{7}{10}x_1 + x_2 \leq 1071 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 1020$ \
$x_1 + \frac{2}{3}x_2 \leq 1203.6$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 229.5$

#### Planteamiento del problema dual de la tercera estrategia

max $g(\lambda) = 1071\lambda_1 + 1020\lambda_2 + 1203.6\lambda_3 + 229.5\lambda_4$ s.a. \
$\frac{7}{10}\lambda_1 + \frac{1}{2}\lambda_2 + \lambda_3 + \frac{1}{10}\lambda_4 \geq -14$ \
$\lambda_1 + \frac{5}{6}\lambda_2 + \frac{2}{3}\lambda_3 + \frac{1}{4}\lambda_4 \geq -9 $

In [17]:
c = np.array([[630*1.7, 600*1.7, 708*1.7, 135*1.7]])
A = np.array([[-7/10, -1/2, -1, -1/10],[-1, -5/6, -2/3, -1/4],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7],[-9],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex')
print(f'La solución a este problema de optimización es λ = {res.x}')
print(f'El valor de la función objetivo del dual evaluada en el valor óptimo es {round(res.fun,3)}')

La solución a este problema de optimización es λ = [8.125  0.     1.3125 0.    ]
El valor de la función objetivo del dual evaluada en el valor óptimo es 10281.6


#### Solución del primal a través de holgura complementaria

A partir de la solución anterior, se evaluán las condiciones de holgura complementaria que permiten encontrar la solución en el problema primal, es decir, las cantidades de bolsas a fabricar de modelo estándar y modelo deluxe. Se obtienen las siguientes restricciones: 

$-\frac{7}{10}(8.125) - \frac{1}{2}(0) - 1(1.3125) -\frac{1}{10}(0) < -7$\
$-7 = -7 \longrightarrow$ Restricción activa 

$-1(8.125) - \frac{5}{6}(0) - \frac{2}{3}(1.3125) -\frac{1}{4}(0) < -9$
$-9 = -9 \longrightarrow$ Restricción activa 

$\lambda_1^*>0\longrightarrow$ Restricción inactiva 

$\lambda_2^*=0\longrightarrow$ Restricción activa 

$\lambda_3^*>0\longrightarrow$ Restricción inactiva 

$\lambda_4^*=0\longrightarrow$ Restricción activa 

De la tercera y quinta restricción, se obtiene respectivamente que:

$-\frac{7}{10}x_1^* - x_2^*=-1071$ \
$-x_1^*-\frac{2}{3}x_2^*=-1203.6$

Finalmente, se resuelve el sistema y se obtiene que se deben fabricar 918 bolsas modelo estándar y 428.4 bolsas modelo deluxe.

## Selección de estrategia

Es más conveniente escoger la tercera estrategia porque de esta forma se obtiene un valor de utilidad total más alto en comparación a las otras dos estrategias. A continuación se encuentra la información resumida:

Utilidad con **primera estrategia**: $9,720

Utilidad con **segunda estrategia**: $9,912

Utilidad con **tercera estrategia**: $10,281.6

# Problem 3

### a) Capacitación de todos los empleados, mejorando todos los tiempos de producción en 10%


In [18]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[(-7/10)*0.9, (-1/2)*0.9, -1*0.9, (-1/10)*0.9],[-1*0.9, (-5/6)*0.9, (-2/3)*0.9, (-1/4)*0.9],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7],[-9],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex', options={"disp": True})
print(f'La solución a este problema de optimización es λ = {res.x}')
print(f'El valor de la función objetivo del dual evaluada en el valor óptimo es {round(res.fun,3)}')

Optimization terminated successfully.
         Current function value: 6720.000000 
         Iterations: 3
La solución a este problema de optimización es λ = [9.02777778 0.         1.45833333 0.        ]
El valor de la función objetivo del dual evaluada en el valor óptimo es 6720.0


#### Solución del primal a través de holgura complementaria

A partir de la solución anterior, se evaluán las condiciones de holgura complementaria que permiten encontrar la solución en el problema primal, es decir, las cantidades de bolsas a fabricar de modelo estándar y modelo deluxe. Se obtienen las siguientes restricciones: 

$-0.63(9.028) - 0.45(0) - 0.9(1.458) -0.09(0) < -7$
$-7 = -7 \longrightarrow$ Restricción activa 

$-0.9(9.028) - 0.75(0) - 0.6(1.458) - 0.225(0) < -9$\
$-9 = -9 \longrightarrow$ Restricción activa 

$\lambda_1^*>0\longrightarrow$ Restricción inactiva 

$\lambda_2^*=0\longrightarrow$ Restricción activa 

$\lambda_3^*>0\longrightarrow$ Restricción inactiva 

$\lambda_4^*=0\longrightarrow$ Restricción activa 

De la tercera y quinta restricción, se obtiene respectivamente que:

$-0.63x_1^* - 0.9x_2^*=-630$ \
$-0.9x_1^*- 0.6x_2^*=-708$

Finalmente, se resuelve el sistema y se obtiene que se deben fabricar 600 bolsas modelo estándar y 280 bolsas modelo deluxe.

In [19]:
Aprim = np.array([[(-7/10)*0.9, -1*0.9],[(-1/2)*0.9,(-5/6)*0.9],[-1*0.9, (-2/3)*0.9],[(-1/10)*0.9,(-1/4)*0.9],[1,0],[0,1]])
bprim = np.array([[-630],[-600],[-708],[-135],[0],[0]])
fprim = np.array([-7, -9])
solucion = res.x
solucion=solucion.reshape(-1,1)
valor_optimo = res.fun

sizeAprim = np.shape(Aprim)
sizebprim = np.shape(bprim)
sizefprim = np.shape(fprim)
sizesolucion = np.shape(solucion)

fdual = bprim[bprim != 0]
sizefdual = np.shape(fdual)

Adual = np.zeros((sizeAprim[1],np.shape(fdual)[0]))
bdual = np.zeros((sizeAprim[1],1))

for i in range(sizeAprim[1]):
    bdual[i] = fprim[i]
    for j in range(sizefdual[0]):
        Aprim_copy = Aprim.copy()
        Aprim_copy = Aprim[:-sizeAprim[1], :]
        Adual[i,j] = Aprim_copy[j][i]
        
sizeAdual = np.shape(Adual)
sizebdual = np.shape(bdual)
sizefdual = np.shape(fdual)

print(f'A del primal: \n {Aprim}')
print(f'b del primal: \n {bprim}')
print(f'f del primal: \n {fprim}')  

print(f'A del dual: \n {Adual}')
print(f'b del dual: \n {bdual}')
print(f'f del dual: \n {fdual}')

sol_primal = np.zeros((sizeAdual[1],sizeAdual[0]))

for i in range(sizeAdual[0]):
    filai = Adual[i]
    valor = np.dot(filai,solucion)
    if valor == bdual[i]:
        pass
    elif valor < bdual[i]:
        fila = sol_primal[i,:]
        for j in range(np.shape(fila)[0]):
            if j == i:
                sol_primal[j+1,i] = 1

bprim_aux = np.zeros((sizesolucion[0],1))

for i in range(sizesolucion[0]):
    if solucion[i] > 0:
        sol_primal[i,:] = Aprim[i,:] 
        bprim_aux[i] = bprim[i]   
    elif solucion[i] == 0:
        sol_primal[i,:] = np.zeros(sizeAprim[1])
        bprim_aux[i] = 0

sol_primal = sol_primal[np.all(sol_primal != 0, axis=1)]
bprim_aux = bprim_aux[np.all(bprim_aux != 0, axis=1)]
        
solx = np.linalg.solve(sol_primal, bprim_aux)
        
print(f'Matriz para solucionar el primal: \n {sol_primal}')
print(f'Vector b para solucionar el primal: \n {bprim_aux}')
print(f'Valores de x: \n {solx}')
print(f'Función objetivo del primal: \n {-1*np.dot(fprim,solx)}')

A del primal: 
 [[-0.63  -0.9  ]
 [-0.45  -0.75 ]
 [-0.9   -0.6  ]
 [-0.09  -0.225]
 [ 1.     0.   ]
 [ 0.     1.   ]]
b del primal: 
 [[-630]
 [-600]
 [-708]
 [-135]
 [   0]
 [   0]]
f del primal: 
 [-7 -9]
A del dual: 
 [[-0.63  -0.45  -0.9   -0.09 ]
 [-0.9   -0.75  -0.6   -0.225]]
b del dual: 
 [[-7.]
 [-9.]]
f del dual: 
 [-630 -600 -708 -135]
Matriz para solucionar el primal: 
 [[-0.63 -0.9 ]
 [-0.9  -0.6 ]]
Vector b para solucionar el primal: 
 [[-630.]
 [-708.]]
Valores de x: 
 [[600.]
 [280.]]
Función objetivo del primal: 
 [6720.]


De acuerdo a este resultado, se obtiene que **se deben fabricar 600 bolsas estándar y 280 bolsas deluxe** para maximizar la utilidad en un valor de **$6,720**. 

### b) Una nueva imágen que le permitiría aumentar la utilidad por bolsa en un 10%

In [20]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[(-7/10), (-1/2), -1, (-1/10)],[-1, (-5/6), (-2/3), (-1/4)],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7*1.1],[-9*1.1],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex', options={"disp": True})
print(f'La solución a este problema de optimización es λ = {res.x}')
print(f'El valor de la función objetivo del dual evaluada en el valor óptimo es {round(res.fun,3)}')

Optimization terminated successfully.
         Current function value: 6652.800000 
         Iterations: 3
La solución a este problema de optimización es λ = [8.9375  0.      1.44375 0.     ]
El valor de la función objetivo del dual evaluada en el valor óptimo es 6652.8


#### Solución del primal a través de holgura complementaria

A partir de la solución anterior, se evaluán las condiciones de holgura complementaria que permiten encontrar la solución en el problema primal, es decir, las cantidades de bolsas a fabricar de modelo estándar y modelo deluxe. Se obtienen las siguientes restricciones: 

$-\frac{7}{10}(8.9375) - \frac{1}{2}(0) - 1(1.44375) -\frac{1}{10}(0) < -7.7$\
$-7.7 = -7.7 \longrightarrow$ Restricción activa 

$-1(8.9375) - \frac{5}{6}(0) - \frac{2}{3}(1.44375) -\frac{1}{4}(0) < -9.9$
$-9.9 = -9.9 \longrightarrow$ Restricción activa 

$\lambda_1^*>0\longrightarrow$ Restricción inactiva 

$\lambda_2^*=0\longrightarrow$ Restricción activa 

$\lambda_3^*>0\longrightarrow$ Restricción inactiva 

$\lambda_4^*=0\longrightarrow$ Restricción activa 

De la tercera y quinta restricción, se obtiene respectivamente que:

$-\frac{7}{10}x_1^* - x_2^*=-630$ \
$-x_1^*-\frac{2}{3}x_2^*=-708$

Finalmente, se resuelve el sistema y se obtiene que se deben fabricar 540 bolsas modelo estándar y 252 bolsas modelo deluxe.

In [21]:
Aprim = np.array([[(-7/10), -1],[(-1/2),(-5/6)],[-1, (-2/3)],[(-1/10),(-1/4)],[1,0],[0,1]])
bprim = np.array([[-630],[-600],[-708],[-135],[0],[0]])
fprim = np.array([-7*1.1, -9*1.1])

solucion = res.x

sizeAprim = np.shape(Aprim)
sizebprim = np.shape(bprim)
sizefprim = np.shape(fprim)
sizesolucion = np.shape(solucion)

fdual = bprim[bprim != 0]
sizefdual = np.shape(fdual)

Adual = np.zeros((sizeAprim[1],np.shape(fdual)[0]))
bdual = np.zeros((sizeAprim[1],1))

for i in range(sizeAprim[1]):
    bdual[i] = fprim[i]
    for j in range(sizefdual[0]):
        Aprim_copy = Aprim.copy()
        Aprim_copy = Aprim[:-sizeAprim[1], :]
        Adual[i,j] = Aprim_copy[j][i]
        
sizeAdual = np.shape(Adual)
sizebdual = np.shape(bdual)
sizefdual = np.shape(fdual)

print(f'A del primal: \n {Aprim}')
print(f'b del primal: \n {bprim}')
print(f'f del primal: \n {fprim}')  

print(f'A del dual: \n {Adual}')
print(f'b del dual: \n {bdual}')
print(f'f del dual: \n {fdual}')

sol_primal = np.zeros((sizeAdual[1],sizeAdual[0]))

for i in range(sizeAdual[0]):
    filai = Adual[i]
    valor = np.dot(filai,solucion)
    if valor == bdual[i]:
        pass
    elif valor < bdual[i]:
        fila = sol_primal[i,:]
        for j in range(np.shape(fila)[0]):
            if j == i:
                sol_primal[j+1,i] = 1

bprim_aux = np.zeros((sizesolucion[0],1))

for i in range(sizesolucion[0]):
    if solucion[i] > 0:
        sol_primal[i,:] = Aprim[i,:] 
        bprim_aux[i] = bprim[i]   
    elif solucion[i] == 0:
        sol_primal[i,:] = np.zeros(sizeAprim[1])
        bprim_aux[i] = 0

sol_primal = sol_primal[np.all(sol_primal != 0, axis=1)]
bprim_aux = bprim_aux[np.all(bprim_aux != 0, axis=1)]
        
solx = np.linalg.solve(sol_primal, bprim_aux)
        
print(f'Matriz para solucionar el primal: \n {sol_primal}')
print(f'Vector b para solucionar el primal: \n {bprim_aux}')
print(f'Valores de x: \n {solx}')
print(f'Función objetivo del primal: \n {-1*np.dot(fprim,solx)}')

A del primal: 
 [[-0.7        -1.        ]
 [-0.5        -0.83333333]
 [-1.         -0.66666667]
 [-0.1        -0.25      ]
 [ 1.          0.        ]
 [ 0.          1.        ]]
b del primal: 
 [[-630]
 [-600]
 [-708]
 [-135]
 [   0]
 [   0]]
f del primal: 
 [-7.7 -9.9]
A del dual: 
 [[-0.7        -0.5        -1.         -0.1       ]
 [-1.         -0.83333333 -0.66666667 -0.25      ]]
b del dual: 
 [[-7.7]
 [-9.9]]
f del dual: 
 [-630 -600 -708 -135]
Matriz para solucionar el primal: 
 [[-0.7        -1.        ]
 [-1.         -0.66666667]]
Vector b para solucionar el primal: 
 [[-630.]
 [-708.]]
Valores de x: 
 [[540.]
 [252.]]
Función objetivo del primal: 
 [6652.8]


De acuerdo a este resultado, se obtiene que **se deben fabricar 540 bolsas estándar y 252 bolsas deluxe** para maximizar la utilidad en un valor de **$6,652.8**. 

## Selección de propuestas

De estas dos últimas propuestas, es más conveniente implementar la ***primera*** propuesta, porque de este modo se obtiene un valor de utilidad mayor.

Utilidad con primera propuesta: **$6,720**

Utilidad con segunda propuesta: **$6,652.8**

# Problem 4

Se define una función que retorna la solución del primal a partir de la información del dual, usando las condiciones de holgura complementaria.

In [22]:
def hcomp(Aprim,bprim,fprim,soldual):
    
    # Se descompone el parametro soldual, para determinar el valor de la solución óptima, y el valor de la función
    # objetivo evaluada en esa solución óptima.
    solucion = soldual[0]
    valor_opt = soldual[1]
    
    fprim = -1*fprim
    
    # Se obtiene el sistema del dual.
    Adual = Aprim.transpose()
    fdual = bprim.transpose()
    bdual = fprim.transpose()
    
    # Se crean listas para guardar los ínidces de las variables que se igualan a cero
    ecuaciones = []
    ceros = []
    
    # Primera condición de hc: Se reemplaza el valor solución en las restricciones del dual. El resultado queda
    # guardado en condicion11.
    condicion11 = Adual.dot(solucion).transpose()
    
    # Se determina si las restricciones de la primera condición de hc son activas o inactivas. 
    # False -> Activa, True -> Inactiva
    restr = [condicion11[k]>bdual[k] for k in range(len(condicion11))]
    
    for j in range(len(restr)):
        # Si la resitricción j es inactiva, significa que la variable xj = 0. Se agrega el índice a una lista.  
        if (restr[j]):
             ceros.append(j)
    
    # Segunda condición de hc: Si las restricciones son inactivas, se toma el índice de la ecuación que va a formar
    # parte de la solución.
    for i in range(len(solucion)):
        if (solucion[i] != 0):
            ecuaciones.append(i)
    
    # Se crea la matriz para solucionar el sistema de ecuaciones resultante de las restricciones activas e inactivas,
    # así como su correspondiente vector b.
    matriz_sol = Aprim.copy()[ecuaciones,:]
    b_sol = bprim.copy()[ecuaciones,:] 
    
    # Se eliminan las columnas del sistema y de la función objetivo donde la xj es nula.
    # Se concatena el sistema resultante con el vector b (matriz aumentada). 
    matriz_sol = np.delete(matriz_sol, ceros, axis=1)
    fprim = np.delete(fprim, ceros, axis=1)
    aumentada = np.c_[matriz_sol, b_sol]
    
    # Se toman los indices de las columnas pivotes, para saber los índices de las ecuaciones que se están tomando.
    reducida , indices = sympy.Matrix(aumentada).rref()
    indices = np.array(indices)
    
    # Se vuelve a definir la matriz solo con las ecuaciones relevantes, así como el vector b. 
    matriz_sol = matriz_sol[indices]
    b_sol = b_sol[indices]
    
    # Se encuentra la solución si hay la misma cantidad de ecuaciones resultantes que valores del vector b.  
    if (len(indices) == len(b_sol)):
        x = np.linalg.inv(matriz_sol) @ b_sol

    # Se determina el valor de la función del primal evaluada en la solución recientemente encontrada.
    val_primal = np.dot(fprim,x)
    
    # Se insertan las variables nulas de nuevo en su posición correspondiente, para indicar cuáles tomaron un valor 
    # igual a cero. 
    x = np.insert(x, ceros, 0)
    
    return val_primal, x

### Función implementada con el problema 2 a):

In [23]:
# Matriz de restricciones
An = np.array([[7/10, 1],[1/2, 5/6],[1, 2/3],[1/10,1/4]])

# Límites superiores de las restricciones
b = np.array([[630],[600],[708],[135]])

# Función objetivo
f = np.array([[-7,-18]])

# Solución del dual. La primera fila corresponde a la solución del problema dual. La segunda fila corresponde al valor
# de la solución evaluada en la función objetivo.
sol = np.array([[0, 0, 0, 72],[9720]])

# Se aplica la función
[fprim, x] = hcomp(An, b, f, sol)

print(f'La solución al problema original de optimización es x = {x}')
print(f'El valor de la función objetivo evaluada en el valor óptimo es {fprim}')

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.

### Función implementada con el problema 2 b):

In [ ]:
# Matriz de restricciones
A = np.array([[7/10, 1],[1/2, 5/6],[1, 2/3],[1/10,1/4]])

# Límites superiores de las restricciones
b = np.array([[630],[600],[708],[135]])

# Función objetivo
f = np.array([[-14,-7]])

# Solución del dual. La primera fila corresponde a la solución del problema dual. La segunda fila corresponde al valor
# de la solución evaluada en la función objetivo.
sol = np.array([[0, 0, 14, 0],[9912]])

# Se aplica la función
[fprim, x] = hcomp(A, b, f, sol)

print(f'La solución al problema original de optimización es x = {x}')
print(f'El valor de la función objetivo evaluada en el valor óptimo es {fprim}')

### Función implementada con el problema 2 c):

In [ ]:
# Matriz de restricciones
A = np.array([[7/10, 1],[1/2, 5/6],[1, 2/3],[1/10,1/4]])

# Límites superiores de las restricciones
b = np.array([[1071],[1020],[1203.6],[229.5]])

# Función objetivo
f = np.array([[-7,-9]])

# Solución del dual. La primera fila corresponde a la solución del problema dual. La segunda fila corresponde al valor
# de la solución evaluada en la función objetivo.
sol = np.array([[8.125, 0, 1.3125, 0],[10281.6]])

# Se aplica la función
[fprim, x] = hcomp(A, b, f, sol)

print(f'La solución al problema original de optimización es x = {x}')
print(f'El valor de la función objetivo evaluada en el valor óptimo es {fprim}')